src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1076,7 +1076,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val sam = samMatchingFunction(tree, pt) // this implies tree.isInstanceOf[Function]
           if (sam.exists) {
             val samTree = adaptToSAM(sam, tree.asInstanceOf[Function], pt, mode)
-            if (samTree ne EmptyTree) return samTree
+            if (samTree ne EmptyTree)
+              return samTree.updateAttachment(SAMFunction(sam.owner))
```

Comment:
When I pass along the sam from SAMFunction to LambdaMetaFactoryCapable, this now fails with a CCE similar to the earlier issue you found:
```
scala> trait T { def foo(a: Any): Any }; trait U extends T { def apply = ??? }
defined trait T
defined trait U

scala>  (((x: Any) => x) : U).foo("")
java.lang.ClassCastException: $$Lambda$1747/323326911 cannot be cast to U
  ... 30 elided
```